api/envoy/extensions/transport_sockets/tls/v3/common.proto

```diff
@@ -254,7 +254,6 @@ message TlsSessionTicketKeys {
 // The plugin instances are defined in the client's bootstrap file.
 // The plugin allows certificates to be fetched/refreshed over the network asynchronously with
 // respect to the TLS handshake.
-// [#not-implemented-hide:]
```

Comment:
I think you also need to remove the `not-implemented-hide` tag on line 352, where this message type is used for the `CertificateValidationContext.ca_certificate_provider_instance` field.  That field is what tells the xDS client (Envoy or gRPC) to use the certificate provider instance to get the root cert used to validate peer certificates.

Reply:
Can I move the tag and leave is unimplemented, in other words we accept the config but do nothing for that.

Reply:
So you're going to support using a CertificateProvider only for the identity cert, not for the root cert?  Is there any reason not to implement both at once?  It seems like we should at least build the infrastructure in the CertificateProvider API for this, even if the one extension that you're actually implementing doesn't support root certs.

Reply:
What do you exactly mean by "build the infrastructure in the CertificateProvider API for this"?

You did remind me that it should be separated to three parts/PRs:
1. implement `certificate_provider_instances` in bootstrap (current PR)
For short it implements certificate_provider_instances management.`CertificateProviderManager` is constructed during config initialization, each entry of certificate_provider_instances in bootstrap will be constructed as a `CertificateProvider` through `CertificateProviderFactory` and then managed by `CertificateProviderManager`
`CertificateProviderFactory` is base factory to create `CertificateProvider` instance
`CertificateProvider` is base class of all instances, having interfaces defined
2. implement `tls_certificate_provider_instance`
3. implement `ca_certificate_provider_instance`

I was working on [issue](https://github.com/envoyproxy/envoy/issues/18928) and [PR](https://github.com/envoyproxy/envoy/pull/19137) before, this is my main proposal but it requires`tls_certificate_provider_instance`. So I prefer complete the item no.1 for first, then item no.2 to support my main proposal, and item no.3 at last.

The title "Implement CertificateProviderPluginInstance mechanism (#19308)" might not be proper here, will modify it to "implement `certificate_provider_instances` in bootstrap"

Reply:
What I meant by "build the infrastructure in the CertificateProvider API for this" is that the `CertificateProvider` interface should have methods for providing both identity certs and root certs.  I think those methods should be present from the get-go on the interface, even if the implementation doesn't work yet, because otherwise anyone else starting to implement their own plugin will not provide both methods, and then it will be harder to add them later.

I'll defer to Envoy folks on the question of how to split up the implementation work into multiple PRs.

Reply:
So it does not conflict with what I proposed. As I mentioned in the first part, we will provide methods in CertificateProvider, that is to say CertificateProvider is base class of all instances, having interfaces defined

Reply:
My concern is that in this PR, the `CertificateProvider` interface does not have any methods at all.  So it seems like the intent is to add the method for providing the identity cert in step 2 and then add the method for providing the root cert in step 3.  If there is not going to be any significant delay between steps 1 and 3, then this seems fine.  But if there is going to be a delay getting to step 3, then that introduces a window in which others might implement new `CertificateProvider`s that have only one of the two methods, which would make it harder to add the second method.

Reply:
Sorry to make you confused. I will provide the methods in `CertificateProvider` later in this PR. It's empty for now because this PR is still WIP, I submit this to ask if the implementation direction is correct.

Reply:
Ah, okay, thanks for the clarification.  If this PR will include both methods on the `CertificateProvider` interface, then that sounds like a great plan from my perspective.